File: frontend/packages/cozeloop/api-schema/src/api/idl/prompt/domain/prompt.ts
export interface Prompt {
  id?: string,
  workspace_id?: string,
  prompt_key?: string,
  prompt_basic?: PromptBasic,
  prompt_draft?: PromptDraft,
  prompt_commit?: PromptCommit,
}
export interface PromptBasic {
  display_name?: string,
  description?: string,
  latest_version?: string,
  created_by?: string,
  updated_by?: string,
  created_at?: string,
  updated_at?: string,
  latest_committed_at?: string,
}
export interface PromptCommit {
  detail?: PromptDetail,
  commit_info?: CommitInfo,
}
export interface CommitInfo {
  version?: string,
  base_version?: string,
  description?: string,
  committed_by?: string,
  committed_at?: string,
}
export interface PromptDraft {
  detail?: PromptDetail,
  draft_info?: DraftInfo,
}
export interface DraftInfo {
  user_id?: string,
  base_version?: string,
  is_modified?: boolean,
  created_at?: string,
  updated_at?: string,
}
export interface PromptDetail {
  prompt_template?: PromptTemplate,
  tools?: Tool[],
  tool_call_config?: ToolCallConfig,
  model_config?: ModelConfig,
}
export interface PromptTemplate {
  template_type?: TemplateType,
  messages?: Message[],
  variable_defs?: VariableDef[],
}
export enum TemplateType {
  Normal = "normal",
}
export interface Tool {
  type?: ToolType,
  function?: Function,
}
export enum ToolType {
  Function = "function",
}
export interface Function {
  name?: string,
  description?: string,
  parameters?: string,
}
export interface ToolCallConfig {
  tool_choice?: ToolChoiceType
}
export enum ToolChoiceType {
  None = "none",
  Auto = "auto",
}
export interface ModelConfig {
  model_id?: string,
  max_tokens?: number,
  temperature?: number,
  top_k?: number,
  top_p?: number,
  presence_penalty?: number,
  frequency_penalty?: number,
  json_mode?: boolean,
}
export interface Message {
  role?: Role,
  reasoning_content?: string,
  content?: string,
  parts?: ContentPart[],
  tool_call_id?: string,
  tool_calls?: ToolCall[],
}
export enum Role {
  System = "system",
  User = "user",
  Assistant = "assistant",
  Tool = "tool",
  Placeholder = "placeholder",
}
export interface ContentPart {
  type?: ContentType,
  text?: string,
  image_url?: ImageURL,
}
export enum ContentType {
  Text = "text",
  ImageURL = "image_url",
}
export interface ImageURL {
  uri?: string,
  url?: string,
}
export interface ToolCall {
  index?: string,
  id?: string,
  type?: ToolType,
  function_call?: FunctionCall,
}
export interface FunctionCall {
  name?: string,
  arguments?: string,
}
export interface VariableDef {
  key?: string,
  desc?: string,
  type?: VariableType,
}
export interface VariableVal {
  key?: string,
  value?: string,
  placeholder_messages?: Message[],
}
export enum VariableType {
  String = "string",
  Placeholder = "placeholder",
}
export interface TokenUsage {
  input_tokens?: string,
  output_tokens?: string,
}
export interface DebugContext {
  debug_core?: DebugCore,
  debug_config?: DebugConfig,
  compare_config?: CompareConfig,
}
export interface DebugCore {
  mock_contexts?: DebugMessage[],
  mock_variables?: VariableVal[],
  mock_tools?: MockTool[],
}
export interface CompareConfig {
  groups?: CompareGroup[]
}
export interface CompareGroup {
  prompt_detail?: PromptDetail,
  debug_core?: DebugCore,
}
export interface DebugMessage {
  role?: Role,
  content?: string,
  reasoning_content?: string,
  parts?: ContentPart[],
  tool_call_id?: string,
  tool_calls?: DebugToolCall[],
  debug_id?: string,
  input_tokens?: string,
  output_tokens?: string,
  cost_ms?: string,
}
export interface DebugToolCall {
  tool_call?: ToolCall,
  mock_response?: string,
  debug_trace_key?: string,
}
export interface MockTool {
  name?: string,
  mock_response?: string,
}
export interface DebugConfig {
  single_step_debug?: boolean
}
export interface DebugLog {
  id?: string,
  prompt_id?: string,
  workspace_id?: string,
  prompt_key?: string,
  version?: string,
  input_tokens?: string,
  output_tokens?: string,
  cost_ms?: string,
  status_code?: number,
  debugged_by?: string,
  debug_id?: string,
  debug_step?: number,
  started_at?: string,
  ended_at?: string,
}
export enum Scenario {
  Default = "default",
  EvalTarget = "eval_target",
}

File: backend/modules/prompt/domain/entity/prompt.go
// Copyright (c) 2025 coze-dev Authors
// SPDX-License-Identifier: Apache-2.0

package entity

import (
	"time"

	"github.com/coze-dev/coze-loop/backend/pkg/lang/mem"
)

type Prompt struct {
	ID           int64         `json:"id"`
	SpaceID      int64         `json:"space_id"`
	PromptKey    string        `json:"prompt_key"`
	PromptBasic  *PromptBasic  `json:"prompt_basic,omitempty"`
	PromptDraft  *PromptDraft  `json:"prompt_draft,omitempty"`
	PromptCommit *PromptCommit `json:"prompt_commit,omitempty"`
}

type PromptDraft struct {
	PromptDetail *PromptDetail `json:"prompt_detail,omitempty"`
	DraftInfo    *DraftInfo    `json:"draft_info,omitempty"`
}

type PromptCommit struct {
	PromptDetail *PromptDetail `json:"prompt_detail,omitempty"`
	CommitInfo   *CommitInfo   `json:"commit_info,omitempty"`
}

type DraftInfo struct {
	UserID      string    `json:"user_id"`
	BaseVersion string    `json:"base_version"`
	IsModified  bool      `json:"is_modified"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

type CommitInfo struct {
	Version     string    `json:"version"`
	BaseVersion string    `json:"base_version"`
	Description string    `json:"description"`
	CommittedBy string    `json:"committed_by"`
	CommittedAt time.Time `json:"committed_at"`
}

func (p *Prompt) Clone() *Prompt {
	if p == nil {
		return nil
	}
	copiedPrompt := &Prompt{}
	_ = mem.DeepCopy(p, copiedPrompt)
	return copiedPrompt
}

func (p *Prompt) CloneDetail() *Prompt {
	if p == nil {
		return nil
	}
	copiedPrompt := p.Clone()
	if copiedPrompt == nil {
		return nil
	}

	copiedPrompt.ID = 0
	copiedPrompt.PromptKey = ""
	copiedPrompt.PromptBasic = nil
	if copiedPrompt.PromptCommit != nil {
		copiedPrompt.PromptCommit.CommitInfo = nil
	}
	if copiedPrompt.PromptDraft != nil {
		copiedPrompt.PromptDraft.DraftInfo = nil
	}

	return copiedPrompt
}

func (p *Prompt) GetVersion() string {
	if p == nil {
		return ""
	}
	var version string
	if p.PromptCommit != nil && p.PromptCommit.CommitInfo != nil {
		version = p.PromptCommit.CommitInfo.Version
	}
	return version
}

func (p *Prompt) GetPromptDetail() *PromptDetail {
	if p == nil {
		return nil
	}
	if p.PromptDraft != nil && p.PromptDraft.PromptDetail != nil {
		return p.PromptDraft.PromptDetail
	}
	if p.PromptCommit != nil && p.PromptCommit.PromptDetail != nil {
		return p.PromptCommit.PromptDetail
	}
	return nil
}

func (p *Prompt) FormatMessages(messages []*Message, variableVals []*VariableVal) (formattedMessages []*Message, err error) {
	if p == nil {
		return nil, nil
	}
	var promptTemplate *PromptTemplate
	if promptDetail := p.GetPromptDetail(); promptDetail != nil {
		promptTemplate = promptDetail.PromptTemplate
	}
	if promptTemplate == nil {
		return nil, nil
	}
	return promptTemplate.formatMessages(messages, variableVals)
}

func (p *Prompt) GetTemplateMessages(messages []*Message) []*Message {
	if p == nil {
		return nil
	}
	var promptTemplate *PromptTemplate
	if promptDetail := p.GetPromptDetail(); promptDetail != nil {
		promptTemplate = promptDetail.PromptTemplate
	}
	if promptTemplate == nil {
		return nil
	}
	return promptTemplate.getTemplateMessages(messages)
}
